src/data/sign.lean

```diff
@@ -202,3 +201,58 @@ def sign_hom : α →*₀ sign_type :=
                mul_neg_of_pos_of_neg, mul_pos] }
 
 end linear_ordered_ring
+
+namespace int
+
+lemma sign_eq_sign (n : ℤ) : n.sign = _root_.sign n :=
+begin
+  obtain ((_ | _) | _) := n,
+  { exact congr_arg coe sign_zero.symm },
+  { exact congr_arg coe (sign_pos $ int.succ_coe_nat_pos _).symm },
+  { exact congr_arg coe (_root_.sign_neg $ neg_succ_lt_zero _).symm }
+end
+end int
+
+open finset nat
+open_locale big_operators
+
+-- TODO: Inlining this yields an app-builder exception
+lemma exists_signed_sum_aux [decidable_eq α] {n : ℕ} (sgn : ℕ → sign_type) (b : α) {f : α → ℤ}
+  ⦃a : α⦄ (g : ℕ → α) (i : ℕ) :
+   (if (range (n - (f a).nat_abs)).piecewise g (λ _, a) i = b then
+       ((range (n - (f a).nat_abs)).piecewise sgn (λ _, sign (f a)) i : ℤ) else 0) =
+    (range (n - (f a).nat_abs)).piecewise (λ j, if g j = b then ↑(sgn j) else 0)
+        (λ j, if a = b then ↑(sign (f a)) else 0) i :=
+by { unfold piecewise, split_ifs; refl }
+
+/-- We can decompose a sum of absolute value less than `n` into a sum of at most `n` signs. -/
+lemma exists_signed_sum [decidable_eq α] [nonempty α] (s : finset α) (n : ℕ) (f : α → ℤ)
+  (hn : ∑ i in s, (f i).nat_abs ≤ n) :
+  ∃ (sgn : ℕ → sign_type) (g : ℕ → α), (∀ i, g i ∉ s → sgn i = 0) ∧
+    ∀ a ∈ s, (∑ i in range n, if g i = a then (sgn i : ℤ) else 0) = f a :=
+begin
```

Comment:
This proof doesn't seem very nice.
Maybe you can use the following and then compose with an injection from `t` into `range n`?
```lean
lemma exists_signed_sum_aux' [decidable_eq α] (s : finset α) (n : ℕ) (f : α → ℤ) :
  ∃ (β : Type u_1) (t : finset β) (sgn : β → sign_type) (g : β → α), (∀ i, g i ∈ s) ∧
    t.card = ∑ i in s, (f i).nat_abs ∧
    ∀ a ∈ s, (∑ i in t, if g i = a then (sgn i : ℤ) else 0) = f a :=
begin
  refine ⟨Σ i : {x // x ∈ s}, ℕ, finset.univ.sigma (λ i, range (f i).nat_abs), λ i, sign (f i.1),
    λ i, i.1, λ i, i.1.prop, _, _⟩,
  { simp [@sum_attach _ _ _ _ (λ i, (f i).nat_abs)] },
  { intros x hx, simp [sum_sigma, hx, ← int.sign_eq_sign, int.sign_mul_nat_abs,
      mul_comm ((f _).nat_abs : ℤ),
      @sum_attach _ _ _ _ (λ i, ∑ j in range (f i).nat_abs, if i = x then (f i).sign else 0)] }
end
```
I assume that `exists_signed_sum_aux'` is not sufficient for the application?

Reply:
I guess transitioning into `range n` will be easier if you define `t` to have cardinality exactly `n` (by adding one element to `\b`)